context.go

```diff
@@ -130,79 +129,61 @@ var testContextData = struct {
 }{service: make(map[string][]byte, 0)}
 
 // Save takes an identifier and an interface. The interface will be network.Marshaled
-// and saved under a filename based on the identifier. An eventual error will be returned.
-// If contextDataPath is non-empty, the destination is a file: it will be created
-// with rw-r----- permissions (0640). If the file already exists, it will be overwritten.
-//
-// The path to the file is chosen as follows:
-//   Mac: ~/Library/Conode/Services
-//   Other Unix: ~/.local/share/conode
-//   Windows: $HOME$\AppData\Local\Conode
-// If the directory doesn't exist, it will be created using rwxr-x---
-// permissions (0750).
-//
-// The path can be overridden with the environmental variable "CONODE_SERVICE_PATH".
+// and saved in the database under the bucket named after the service name.
+// The database must be created by the server prior to using this function.
 func (c *Context) Save(id string, data interface{}) error {
 	buf, err := network.Marshal(data)
 	if err != nil {
 		return err
 	}
-	fname := c.absFilename(id)
-	if getContextDataPath() == "" {
-		testContextData.Lock()
-		testContextData.service[fname] = buf
-		testContextData.Unlock()
-		return nil
-	}
-
-	return ioutil.WriteFile(fname, buf, 0640)
+	bucketName := ServiceFactory.Name(c.ServiceID())
+	return c.manager.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucketName))
+		return b.Put([]byte(id), buf)
+	})
 }
 
 // Load takes an id and returns the network.Unmarshaled data. If an error
-// occurs, the data is nil. See Save() for where the files are saved.
-//
-// If no data is found, it returns an error.
+// occurs, the data is nil. If no data is found, it returns an error.
 func (c *Context) Load(id string) (interface{}, error) {
 	var buf []byte
-	if getContextDataPath() == "" {
-		var ok bool
-		testContextData.Lock()
-		buf, ok = testContextData.service[c.absFilename(id)]
-		testContextData.Unlock()
-		if !ok {
-			return nil, errors.New("this entry doesn't exist")
-		}
-	} else {
-		var err error
-		buf, err = ioutil.ReadFile(c.absFilename(id))
-		if err != nil {
-			return nil, err
-		}
+	bucketName := ServiceFactory.Name(c.ServiceID())
+	c.manager.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucketName))
+		v := b.Get([]byte(id))
+		buf = make([]byte, len(v))
+		copy(buf, v)
+		return nil
+	})
+
+	if len(buf) == 0 {
+		return nil, errors.New("Key does not exist: " + id)
 	}
+
 	_, ret, err := network.Unmarshal(buf, c.server.suite)
 	return ret, err
 }
 
-// DataAvailable checks if any data is stored either in a file or in the
-// contextData map.
+// DataAvailable checks if any data exists under the key `id`
 func (c *Context) DataAvailable(id string) bool {
-	if getContextDataPath() == "" {
-		testContextData.Lock()
-		_, ok := testContextData.service[c.absFilename(id)]
-		testContextData.Unlock()
-		return ok
-	}
-	_, err := os.Stat(c.absFilename(id))
-	return !os.IsNotExist(err)
+	available := false
```

Comment:
Kind of hard to see why DataAvailable is such a useful part of the API. This work is already done, but you should never be afraid of saying, "instead of changing this code, I'm just going to remove it and adapt the 2 places that call it". For example here, Load() could be documented to return nil, nil when no data is available, and boom, you're done.

Reply:
Good point. Actually, there are only two places where `DataAvailable` is used (in cothority)
```
identity/service.go
697:	if !s.DataAvailable("storage") {

pop/service/service.go
854:	if !s.DataAvailable("storage") {
```
In both of these cases `s.Load` immediately follows `s.DataAvailable`. I'd be clearer and simpler to remove this function, it'll also be closer to the functionality in boltdb.

Reply:
IIRC I used `DataAvailable` because I can store a `nil` data - and having `nil` data is different from having no data. But perhaps that's just my memory playing me tricks...